### DISCUSSION TITLE
Abandoned payment processor setup with Stripe free trial

### DISCUSSION THREAD

**Discussion by computer-smile** [pay-rails/pay#570] [Q&A]
I'm creating my Stripe free trial experience. Here is what I have so far. 

1. When a user signs up for the first time, create a fake local subscription.
```ruby
 def set_free_trial
  time = 27.days.from_now
  self.set_payment_processor :fake_processor, allow_fake: true
  self.payment_processor.subscribe(trial_ends_at: time, ends_at: time, metadata: {trial: true})
 end
```
Use this to give them access to features. 

2. When they are ready to activate a paid subscription via stripe, present them with a button that redirects to stripe checkout. Based the recommended strategy here https://github.com/pay-rails/pay/blob/master/docs/stripe/1_overview.md#how-to-use-stripe-checkout-with-pay

```ruby

# app/views/settings.html.erb
<%= button_to "Set up #{plan.capitalize} subscription",checkout_path %>


# app/controllers/checkout

  def checkout
   # this line will create the stripe payment processor and set it as the default. https://github.com/pay-rails/pay/blob/d55823bf5a746b5a71240ba7f6d813cf9b712ede/lib/pay/attributes.rb#L32
    current_user.set_payment_processor :stripe

    @checkout_session = current_user.payment_processor.checkout(
      mode: "subscription",
      line_items: "price_87wuw....."
      )
      redirect_to @checkout_session.url, allow_other_host: true, status: :see_other
  end
```
Just clicking the button from `app/views/settings.html.erb` creates the stripe payment processor and makes it the default. 

3. I then listen for the `stripe.customer.subscription.created` webhook to cancel the trial `Pay::Subscription` created when the user signs up. 
4. So after completing checkout the user has:
- 1 cancelled `Pay::Subscription` that was created for the trial
- 1 `default: false` `Pay::Customer` with `fake_processor` created for the trial
- 1 active `Pay::Subscription` to use moving forward
- 1  `default:true` `Pay::Customer` to be used moving forward

This feels ok as a strategy for free trials with Stripe. It's loosely recommended here https://github.com/pay-rails/pay/discussions/563#discussioncomment-2334577. #563 

2 questions:
1. Has anyone found a better strategy they like? Possible I may be misunderstanding some of the Pay architecture. 

2. If a user abandons checkout in step 2 a `Pay:Customer` will have been created just from clicking my button in `app/views/settings.html.erb` via `current_user.set_payment_processor :stripe`. So, when they abandon they'll have a new default payment processor (stripe) with no active subscription. And the fake trial and its helper methods will no longer work. So they can get themselves in trouble. 

Is this the wrong approach? Did I misunderstand the docs for this action/flow?

Thanks for taking a look. Any recommendation or insight is welcome.

**Comment by kjf** [comment]
@computer-smile Curious to know if you found a flow that works for you?
